mail/api_test.py

```diff
@@ -576,3 +589,146 @@ def test_send_to_course_team_error(self, raise_for_status, mock_post):
+            sender_name=self.automatic_email.sender_name,
+        )
+
+    def test_no_matching_query(self):
+        """If there are no queries matching percolate we should do nothing"""
+        with patch(
+            'mail.api.search_percolate_queries', autospec=True, return_value=[],
+        ) as mock_search_queries, patch('mail.api.MailgunClient') as mock_mailgun:
+            send_automatic_emails(self.program_enrollment_unsent)
+
+        mock_search_queries.assert_called_with(self.program_enrollment_unsent.id)
+        assert mock_mailgun.send_individual_email.called is False
+
+    def test_not_enabled(self):
+        """If the automatic email is not enabled we should do nothing"""
+        with patch(
+            'mail.api.search_percolate_queries', autospec=True, return_value=[self.percolate_query_disabled],
+        ) as mock_search_queries, patch('mail.api.MailgunClient') as mock_mailgun:
+            send_automatic_emails(self.program_enrollment_unsent)
+
+        mock_search_queries.assert_called_with(self.program_enrollment_unsent.id)
+        assert mock_mailgun.send_individual_email.called is False
+
+    def test_already_sent(self):
+        """If a user was already sent email we should not send it again"""
+        with patch(
+            'mail.api.search_percolate_queries', autospec=True, return_value=[self.percolate_query],
+        ) as mock_search_queries, patch('mail.api.MailgunClient') as mock_mailgun:
+            send_automatic_emails(self.program_enrollment_sent)
+
+        mock_search_queries.assert_called_with(self.program_enrollment_sent.id)
+        assert mock_mailgun.send_individual_email.called is False
+
+    def test_failed_send(self):
+        """If we fail to send email to the first user we should still send it to the second"""
+
+        with mute_signals(post_save):
+            new_percolate = PercolateQuery.objects.create(
+                query="new query",
+                original_query="original query",
+            )
+        new_automatic = AutomaticEmail.objects.create(
+            enabled=True,
+            email_subject='new subject',
+            email_body='new body',
+            sender_name='new Sender name',
+            query=new_percolate,
+        )
+
+        count = 0
+
+        def _send_individual_email(*args, **kwargs):  # pylint: disable=unused-argument
+            """Helper method to raise an error the first time it's called, but not the second"""
+            nonlocal count
+            count += 1
+            if count == 1:
+                raise KeyError()
+
+        with patch('mail.api.search_percolate_queries', autospec=True, return_value=[
+            self.percolate_query, new_automatic.query
+        ]) as mock_search_queries, patch(
+            'mail.api.MailgunClient', send_individual_email=_send_individual_email
+        ) as mock_mailgun:
+            mock_mailgun.send_individual_email = _send_individual_email
```

Comment:
2 things:

1. Is it necessary to set `send_individual_email` on the mock object in the `patch()` _and_ in the body of the `with` block?
1. I think you can accomplish this error-first-then-succeed behavior using `side_effect`: https://docs.python.org/3/library/unittest.mock.html#quick-guide. If you give `side_effect` a list, it returns successive elements of that list. You might be able to do something like:

```python
patch(
    'mail.api.MailgunClient', 
    send_individual_email=Mock(side_effect=[Exception(), None])
)
```